### PR TITLE
Externalise Runner Manager

### PR DESCRIPTION
**Pull request by cbartz**
Applicable spec: n/a

### Overview

Externalise the refactored runner interface and dependent modules into a separate python package (https://github.com/canonical/github-runner-manager/pull/2)

### Rationale

A step forward in having a separate Runner Manager application that is independent of the Charm and Charm State. 

### Juju Events Changes

n/a

### Module Changes

Move any modules that the refactored runner interface depends on.

### Library Changes

n/a

### Checklist

- [x] The [charm style guide](https://juju.is/docs/sdk/styleguide) was applied
- [x] The [contributing guide](https://github.com/canonical/is-charms-contributing-guide) was applied
- [x] The changes are compliant with [ISD054 - Managing Charm Complexity](https://discourse.charmhub.io/t/specification-isd014-managing-charm-complexity/11619)
- [x] The documentation is generated using `src-docs`
- [x] The documentation for charmhub is updated.
- [x] The PR is tagged with appropriate label (`urgent`, `trivial`, `complex`)

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -25,7 +25,7 @@ omit = [
 ]
 
 [tool.coverage.report]
-fail_under = 83
+fail_under = 83  # TODO: increase to 85 to match team standard
```
Comment (jdkandersson) [inline pyproject.toml:28]: perhaps make a Jira story for this

File: requirements.txt [hunk 2]
```diff
@@ -14,3 +14,4 @@ PyYAML ==6.0.*
 pyOpenSSL==24.2.1
 kombu==5.4.0
 pymongo==4.8.0
+github_runner_manager @ git+https://github.com/canonical/github-runner-manager.git@initial-app
```
Comment (jdkandersson) [inline requirements.txt:17]: new line?

File: src-docs/charm_state.md
Comment (jdkandersson) [inline src-docs/charm_state.md:1]: Check this

File: src-docs/errors.md
Comment (jdkandersson) [inline src-docs/errors.md:1]: Perhaps just import from the package and delete the definitions from here

File: src-docs/github_client.md
Comment (jdkandersson) [inline src-docs/github_client.md:1]: import from the package and make this class a child adding any additional required methods

**Review by jdkandersson — COMMENTED** [review]
* for utilities let's import them from the package rather than duplicating
* For repo policy compliance lets also import rather than duplicate

File: src-docs/charm_state.md
Comment (cbartz) [reply src-docs/charm_state.md:1]: There was also `charm_state.py.md`, for some reason changes in this PR caused src-docs to create `charm_state.md`. I have removed `charm_state.py.md`.

File: src/utilities.py [hunk 3]
```diff
@@ -3,14 +3,18 @@
...
-from typing import Any, Callable, Optional, Sequence, Type, TypeVar
-
+from typing import Any, Optional, Sequence, TypeVar
+
+# we import the functions from the utilities module, these are used in the charm
+from github_runner_manager.utilities import retry  # noqa: F401  pylint: disable=unused-import
+from github_runner_manager.utilities import (  # noqa: F401  pylint: disable=unused-import
+    secure_run_subprocess,
+    set_env_var,
+)
```
Comment (yanksyoon) [inline src/utilities.py:17]: I believe this is done to make the code migration easier for this PR? (Should we import them where they are required by the module in follow ups?)

File: tests/integration/test_runner_manager_openstack.py [hunk 4]
```diff
@@ -1,5 +1,5 @@
-# Copyright 2024 Canonical Ltd.
-# See LICENSE file for licensing details.
+#  Copyright 2024 Canonical Ltd.
+#  See LICENSE file for licensing details.
```
Comment (yanksyoon) [inline tests/integration/test_runner_manager_openstack.py:2]: Just out of curiosity, is the extra space intended?

Comment (yanksyoon) [reply tests/integration/test_runner_manager_openstack.py:2]: (as in this is what the linter/auto formatters prefer)

**Review by jdkandersson — DISMISSED** [review]

**Review by yanksyoon — DISMISSED** [review]
This seems like a lot of effort, thanks for the changes!

File: src/utilities.py (hunk 3 above)
Comment (cbartz) [reply src/utilities.py:17]: correct. @jdkandersson suggested this approach. We can change this in follow-ups.

**Review by yanksyoon — APPROVED** [review]

File: tests/integration/test_runner_manager_openstack.py (hunk 4 above)
Comment (cbartz) [reply tests/integration/test_runner_manager_openstack.py:2]: I moved the file to the github-runner-manager and checked it again here. I am not sure what is causing the extra space.

**Review by jdkandersson — APPROVED** [review]

File: src/utilities.py (hunk 3 above)
Comment (jdkandersson) [reply src/utilities.py:17]: It is a good intermediate step and then we can clean it up in follow up steps

**Pull request merged**